sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -186,6 +186,13 @@ abstract class Optimizer(catalogManager: CatalogManager)
     // plan may contain nodes that do not report stats. Anything that uses stats must run after
     // this batch.
     Batch("Early Filter and Projection Push-Down", Once, earlyScanPushDownRules: _*) :+
+    // This batch contains rules that should be applied to writes early. For example,
+    // we have to construct a logical write early so that we can inject needed repartition/sort
+    // operators to satisfy data source distribution and ordering requirements.
+    // Expression optimizations must be run before this batch so that we have optimal
+    // expressions when we construct writes. At the same time, rules that dedup repartition and
+    // sort operators must by run afterwards.
+    Batch("Early Writes", Once, earlyWriteRules: _*) :+
```

Comment:
You can avoid the need to make the rules `Idempotent` by using `FixedPoint(1)` instead of `Once`. There are a few existing rules that do it, like the "Join Reorder" just below.

Reply:
Fixed.